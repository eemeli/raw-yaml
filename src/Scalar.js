import Node from './Node'
import Range from './Range'

export default class Scalar extends Node {
  static endOfDoubleQuote (src, offset) {
    let ch = src[offset]
    while (ch && ch !== '"') {
      offset += (ch === '\\') ? 2 : 1
      ch = src[offset]
    }
    return offset + 1
  }

  static endOfSingleQuote (src, offset) {
    let ch = src[offset]
    while (ch) {
      if (ch === "'") {
        if (src[offset + 1] !== "'") break
        ch = src[offset += 2]
      } else {
        ch = src[offset += 1]
      }
    }
    return offset + 1
  }

  /**
   * Parses a scalar value from the source
   *
   * Accepted forms are:
   * ```
   * @alias
   *
   * "double \"quoted\" string"
   *
   * 'single ''quoted'' string'
   * ```
   * where both forms of quoted string may extend across multiple rows
   * regardless of indentation, and each form may be followed by a
   * whitespace-separated #comment.
   *
   * @param {ParseContext} context
   * @param {number} start - Index of first character
   * @returns {number} - Index of the character after this scalar
   */
  parse (context, start) {
    this.context = context
    const { src } = context
    let offset
    switch (this.type) {
      case Node.Type.ALIAS:
        start += 1
        offset = Node.endOfIdentifier(src, start)
        break
      case Node.Type.QUOTE_DOUBLE:
        offset = Scalar.endOfDoubleQuote(src, start + 1)
        break
      case Node.Type.QUOTE_SINGLE:
        offset = Scalar.endOfSingleQuote(src, start + 1)
        break
      default:
        this.error = new Error(`Unknown node type: ${JSON.stringify(this.type)}`)
        offset = start + 1
    }
    this.valueRange = new Range(start, offset)
    offset = Node.endOfWhiteSpace(src, offset)
    offset = this.parseComment(offset)
    trace: this.type, { valueRange: this.valueRange, comment: this.comment }, JSON.stringify(this.rawValue)
    return offset
  }
}
