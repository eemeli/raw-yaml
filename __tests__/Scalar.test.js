import Scalar from '../src/Scalar'

describe('internals', () => {
  test('constructor', () => {
    const src = 'src'
    const s = new Scalar(src)
    expect(s.src).toBe(src)
  })
  test('endIndent', () => {
    const src = '  - value\n- other\n'
    const s = new Scalar(src)
    const in1 = s.endIndent(0)
    expect(in1).toBe(2)
    const offset = src.indexOf('\n') + 1
    const in2 = s.endIndent(offset)
    expect(in2).toBe(offset)
  })
})

const testScalarParse = ({ pre, post, str, comment, expected }) => {
  let body = str
  if (comment) {
    const lines = body.split('\n')
    lines[0] += ` #${comment}`
    body = lines.join('\n')
  }
  const scalar = new Scalar(pre + body + post)
  const indent = scalar.endIndent(0)
  const end = scalar.parse(pre.length, indent, false)
  const expectedEnd = scalar.endWhiteSpace(pre.length + body.length)
  expect(scalar.rawValue).toBe(expected || str)
  expect(end).toBe(expectedEnd)
  if (comment) expect(scalar.comment).toBe(comment)
  expect(scalar).toMatchSnapshot()
}

const commonTests = {
  'bare': { pre: '', post: '' },
  'newline before & after': { pre: '\n', post: '\n' },
  'complex mapping key': { pre: '? ', post: ' : ' },
  'seq value': { pre: '- ', post: '\n- ' },
  'indented block': { pre: '    - ', post: '\n  x' },
  'flow seq value': { pre: '[ ', post: ' ]' },
  'with comment': { pre: '\n  ', comment: 'comment # here!', post: '\n' }
}

describe('parse "quoted"', () => {
  for (const name in commonTests) {
    const props = Object.assign({ str: '"value"' }, commonTests[name])
    test(name, () => testScalarParse(props))
  }
  test('without spaces', () => testScalarParse({ pre: '{', str: '"value"', post: ',' }))
  test('multi-line', () => testScalarParse({ pre: '\n', str: '"value\nwith\nmore lines"', post: '\n' }))
  test('escaped', () => testScalarParse({ pre: '\n', str: '" #value\\\\\nwith \\"more\\" lines\\""', post: '\n' }))
})

describe("parse 'quoted'", () => {
  for (const name in commonTests) {
    const props = Object.assign({ str: "'value'" }, commonTests[name])
    test(name, () => testScalarParse(props))
  }
  test('without spaces', () => testScalarParse({ pre: '{', str: "'value'", post: ',' }))
  test('multi-line', () => testScalarParse({ pre: '\n', str: "'value\nwith\nmore lines'", post: '\n' }))
  test('escaped', () => testScalarParse({ pre: '\n', str: "' #value\nwith ''more'' lines'''", post: '\n' }))
})

describe("parse *alias", () => {
  for (const name in commonTests) {
    const props = Object.assign({ str: '*alias', expected: 'alias' }, commonTests[name])
    test(name, () => testScalarParse(props))
  }
})

describe("parse >block", () => {
  const block = '      #multiline\n  \n      \tblock'
  for (const name in commonTests) {
    const props = Object.assign({ str: `>\n${block}`, expected: block }, commonTests[name])
    if (props.post && props.post[0] !== '\n') {
      props.pre = `  ${props.pre}`
      props.post = props.post.replace(/^\s?/, '\n')
    }
    test(name, () => testScalarParse(props))
  }
  test('literal with header', () => testScalarParse({ pre: '\n- ', str: `|+2\n${block}`, expected: block, post: '\n' }))
})
