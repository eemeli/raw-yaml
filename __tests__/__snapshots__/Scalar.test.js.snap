// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse "quoted" bare 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 7,
    "start": 0,
  },
  "src": "\\"value\\"",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 7,
    "start": 0,
  },
}
`;

exports[`parse "quoted" complex mapping key 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 10,
    "start": 2,
  },
  "src": "? \\"value\\" : ",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse "quoted" escaped 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 34,
    "start": 1,
  },
  "src": "
\\" #value\\\\\\\\
with \\\\\\"more\\\\\\" lines\\\\\\"\\"
",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 34,
    "start": 1,
  },
}
`;

exports[`parse "quoted" flow seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 10,
    "start": 2,
  },
  "src": "[ \\"value\\" ]",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse "quoted" indented block 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 13,
    "start": 6,
  },
  "src": "    - \\"value\\"
  x",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 13,
    "start": 6,
  },
}
`;

exports[`parse "quoted" multi-line 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 24,
    "start": 1,
  },
  "src": "
\\"value
with
more lines\\"
",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 24,
    "start": 1,
  },
}
`;

exports[`parse "quoted" newline before & after 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 8,
    "start": 1,
  },
  "src": "
\\"value\\"
",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 8,
    "start": 1,
  },
}
`;

exports[`parse "quoted" seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 9,
    "start": 2,
  },
  "src": "- \\"value\\"
- ",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse "quoted" with comment 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": Range {
    "end": 27,
    "start": 12,
  },
  "nodeRange": Range {
    "end": 27,
    "start": 3,
  },
  "src": "
  \\"value\\" #comment # here!
",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 10,
    "start": 3,
  },
}
`;

exports[`parse "quoted" without spaces 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 8,
    "start": 1,
  },
  "src": "{\\"value\\",",
  "tag": null,
  "type": "DOUBLE",
  "valueRange": Range {
    "end": 8,
    "start": 1,
  },
}
`;

exports[`parse 'quoted' bare 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 7,
    "start": 0,
  },
  "src": "'value'",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 7,
    "start": 0,
  },
}
`;

exports[`parse 'quoted' complex mapping key 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 10,
    "start": 2,
  },
  "src": "? 'value' : ",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse 'quoted' escaped 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 32,
    "start": 1,
  },
  "src": "
' #value
with ''more'' lines'''
",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 32,
    "start": 1,
  },
}
`;

exports[`parse 'quoted' flow seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 10,
    "start": 2,
  },
  "src": "[ 'value' ]",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse 'quoted' indented block 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 13,
    "start": 6,
  },
  "src": "    - 'value'
  x",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 13,
    "start": 6,
  },
}
`;

exports[`parse 'quoted' multi-line 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 24,
    "start": 1,
  },
  "src": "
'value
with
more lines'
",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 24,
    "start": 1,
  },
}
`;

exports[`parse 'quoted' newline before & after 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 8,
    "start": 1,
  },
  "src": "
'value'
",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 8,
    "start": 1,
  },
}
`;

exports[`parse 'quoted' seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 9,
    "start": 2,
  },
  "src": "- 'value'
- ",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 9,
    "start": 2,
  },
}
`;

exports[`parse 'quoted' with comment 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": Range {
    "end": 27,
    "start": 12,
  },
  "nodeRange": Range {
    "end": 27,
    "start": 3,
  },
  "src": "
  'value' #comment # here!
",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 10,
    "start": 3,
  },
}
`;

exports[`parse 'quoted' without spaces 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 8,
    "start": 1,
  },
  "src": "{'value',",
  "tag": null,
  "type": "SINGLE",
  "valueRange": Range {
    "end": 8,
    "start": 1,
  },
}
`;

exports[`parse *alias bare 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 6,
    "start": 0,
  },
  "src": "*alias",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 6,
    "start": 1,
  },
}
`;

exports[`parse *alias complex mapping key 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 9,
    "start": 2,
  },
  "src": "? *alias : ",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 8,
    "start": 3,
  },
}
`;

exports[`parse *alias flow seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 9,
    "start": 2,
  },
  "src": "[ *alias ]",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 8,
    "start": 3,
  },
}
`;

exports[`parse *alias indented block 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 12,
    "start": 6,
  },
  "src": "    - *alias
  x",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 12,
    "start": 7,
  },
}
`;

exports[`parse *alias newline before & after 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 7,
    "start": 1,
  },
  "src": "
*alias
",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 7,
    "start": 2,
  },
}
`;

exports[`parse *alias seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": null,
  "nodeRange": Range {
    "end": 8,
    "start": 2,
  },
  "src": "- *alias
- ",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 8,
    "start": 3,
  },
}
`;

exports[`parse *alias with comment 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": null,
  "commentRange": Range {
    "end": 26,
    "start": 11,
  },
  "nodeRange": Range {
    "end": 26,
    "start": 3,
  },
  "src": "
  *alias #comment # here!
",
  "tag": null,
  "type": "ALIAS",
  "valueRange": Range {
    "end": 9,
    "start": 4,
  },
}
`;

exports[`parse >block bare 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 34,
    "start": 0,
  },
  "src": ">
      #multiline
  
      	block",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 34,
    "start": 2,
  },
}
`;

exports[`parse >block complex mapping key 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 38,
    "start": 4,
  },
  "src": "  ? >
      #multiline
  
      	block
: ",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 38,
    "start": 6,
  },
}
`;

exports[`parse >block flow seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 38,
    "start": 4,
  },
  "src": "  [ >
      #multiline
  
      	block
]",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 38,
    "start": 6,
  },
}
`;

exports[`parse >block indented block 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 40,
    "start": 6,
  },
  "src": "    - >
      #multiline
  
      	block
  x",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 40,
    "start": 8,
  },
}
`;

exports[`parse >block literal with header 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": "|+2",
  "commentRange": null,
  "nodeRange": Range {
    "end": 39,
    "start": 3,
  },
  "src": "
- |+2
      #multiline
  
      	block
",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 39,
    "start": 7,
  },
}
`;

exports[`parse >block newline before & after 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 35,
    "start": 1,
  },
  "src": "
>
      #multiline
  
      	block
",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 35,
    "start": 3,
  },
}
`;

exports[`parse >block seq value 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": null,
  "nodeRange": Range {
    "end": 36,
    "start": 2,
  },
  "src": "- >
      #multiline
  
      	block
- ",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 36,
    "start": 4,
  },
}
`;

exports[`parse >block with comment 1`] = `
Scalar {
  "anchor": null,
  "blockStyle": ">",
  "commentRange": Range {
    "end": 21,
    "start": 6,
  },
  "nodeRange": Range {
    "end": 54,
    "start": 3,
  },
  "src": "
  > #comment # here!
      #multiline
  
      	block
",
  "tag": null,
  "type": "BLOCK",
  "valueRange": Range {
    "end": 54,
    "start": 22,
  },
}
`;
